### ISSUE TITLE
ViewContainerRefs createEmbeddedView method should allow to specify injector

### ISSUE THREAD

**Issue by kemsky** [angular/angular#14935]
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[x] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
Currently template references get injector reference from component where templates are defined, later when template instance is created using `createEmbeddedView` all child components get injector from template reference, but not from actual component where template view is created. See plunker from #14842.

**Expected behavior**
It should work similar to `createComponent`, where you could specify injector if needed.

**What is the motivation / use case for changing the behavior?**
You can not use any injection feature until this behavior exists. Impossible to pass/override configuration, provide services and so on. It is very limited solution, which is desired only in some cases, also this might be the reason for #14842. 

Also this behavior is different from `ng-content`, which uses correct injector.

* **Angular version:** 2.4.9 - 4.0.0 RC2

* **Browser:** [all ]
 
* **Language:** [all]

**Comment by cgatian** [comment]
@kemsky Have you found a workaround for this?  I also thought this was an strange limitation.

**Comment by tolemac** [comment]
It would be great to be implemented.

@cgatian, when I need inject parents on childen and I'm using createEmbeddedView, I'm using the context to pass the parent... after that I have to put it on the children manually ...

**Comment by cgatian** [comment]
@tolemac thats what I was afraid of.  Just to give a little context to what we're talking about.

The only way to get services into the `child-test` component is to add them to the `context` property of when using `createEmbeddedView` and to then pass reference to the component by utilizing an `@Input` property on the `child-test` component.  Ideally when you should be able allow the `EmbeddedView` to use the Injector of the `ViewContainerRef`.  

```
 <container-component>
              <ng-template let-title="title" let-userService="userService">
                <child-test [userService]="userService"></child-test>
              </ng-template>
</container-component>
```

**Comment by kemsky** [comment]
`TemplateRef` stores reference to declaring view via private `_parentView` field, so dirty way is to patch this field using current view. See https://github.com/angular/angular/blob/ce1d7c4a6e46de00fc0c4aae0d046de34d6a8395/packages/core/src/view/refs.ts#L315

**Comment by cgatian** [comment]
@kemsky Unfortunately, I don't think it's that simple.  I tried updating and different errors occur that are way too deep in the compiler process.  

Attempting to do the following, doesn't work.
```
(this.template as any)._parentView= (this.viewContainerRef as any)._view;
```

**Comment by Krustie101** [comment]
Debugging the afterViewChecked and afterViewInit methods of a component inside a template, it shows that the lifecycle / change detection events for that component are triggered from the component where the template is defined, not the component where the embedded view is added to (to be expected after reading this issue).  A proper solution should also hook the lifecycle / change detection events of components  inside a template in the lifecycle / change detection events of the component where the embeded view is added. 

In my opinion, it is puzzling why this is even the default behavior. It shows to me that templates are designed to be used near to the location where they are defined. Too bad.

In the docs of templateRef you will find this excerpt 

> Typically new Embedded Views are attached to the View Container of this location, but in advanced use-cases, the View can be attached to a different container while keeping the data-binding and injection context from the original location.

So the current behavior is in line with the documentation.

**Comment by kemsky** [comment]
Honestly i do not know why this behavior was chosen as default, perhaps team was trying to make it simple , but it turned out to be pretty useless for any advanced case, which is disappointing.

**Comment by gund** [comment]
I recently hit this very limitation.

My use-case was related to building a framework for dynamic forms aka you give it JSON config and it will render all the controls run all validations extensions etc.

And so I've got a bunch of components and one major compoenet `my-form` - which is the entry point for any form. It can either take JSON input and render components or it can take components as projected nodes.

And because `my-form` component provides a bunch of stuff in injector for dynamic components (via `ngComponentOutletInjector`), all projected components can never receive/query those things because they are effectively outside.

This is pretty limiting situation and I guess there are 2 options to solve it:
1. Purely by framework: make projected content's injector be were it is projected and not declared
2. Give a hook for developers: Allow developers to hook into injector's resolution mechanism and return custom stuff

The first solution, of course, would be ideal - but might be tricky to implement.
But the second one is really easy yet powerful enough to fulfill most of use-cases since projected content already has access to components injector being projected into, so it's possible to hook in that component's injector.

P.S. In order to better understand what limitation I'm talking about [here is the repo](https://github.com/gund/test-ng-content-projection) that demonstrates it (you can follow link to just see output).

**Comment by dirkluijk** [comment]
I have this use case:

```html
<div *appFoo appBar></div>

<!-- same as: -->

<ng-template appFoo>
  <div appBar></div>
</ng-template>
```
I want `appFoo` to provide a service to `appBar`. I was surprised that this does not work!

**Comment by dirkluijk** [comment]
Can someone explain why the `createComponent()` method supports accepting a custom `Injector` and `createEmbeddedView()` not? Is there a reason, and if not, are there any objections against it?

**Comment by simeyla** [comment]
For my use-case the solution would have to include a way to 're-parent' a control. I use a customized `*ngTemplateOutlet` in order to detach views and rearrange them based on the screen width. I don't want to recreate them every time since they contain state and/or videos.

(Excuse my probable misuse of the correct terminology here but...) :
While it was annoying to find out that they couldn't automatically inherit from the parent injector, it was more confusing wondering what should even happen if a control is re-parented to be underneath another 'injector'.

For my case, each 'box' is contained within a 'group', and when the screen is resized new 'groups' are created and the existing boxes are moved into them. So if I were to consume an injected 'group service' from the constructor of a 'box' how would I change it later? In this case does injection even really make sense - and I didn't have any choice but to use an @Input parameter to update the new 'group'. 

Not sure if this scenario is at all related to why you can't do this currently, but sometimes we think we know what we want (!) but it doesn't quite make sense.

**Comment by vsDizzy** [comment]
I've got into this trying to switch between horizontal and vertical angular material stepper, leaving content the same.

**Comment by schippie** [comment]
I know this is considered an old task but I've also ran into the same issue as the users above. The `createEmbeddedView` would ideally allow us to specify a different Injector for the template that is being embedded. 

Our use case for this is:

Have the `dataSource` used within the angular material `DataTable` be something components which are situation-ed around the table to have access to without having to have every component have an input. It allows us to have the `DataSource` be something we can override on multiple levels. Module, Page and in this situation on a sub-section of the page without the components needing any additional logic. 
Since for 9 out of the 10 lists we build we can specify the DataSource which pulls it's data from a ngrx store directly near the model as it isn't any different only the bucket (sub-store) from which data is retrieved.

**Comment by blemaire** [comment]
Same issue here, for a tabs component which includes a horizontally scrolling strips-bar

The stips-bar is an abstraction to create a horizontal bar of elements with automatically enabled button to scroll left/right when the buttons overflow: 

![image](https://user-images.githubusercontent.com/7140322/62599573-df520d80-b8ec-11e9-95ab-b4932061a224.png)

To function correctly, the bar has a main component. StripsBarComponent and StripBarOptionDirective which injects the StripBarComponent:

```ts
export class stripBarOptionDirective implements AfterContentInit {
  constructor(@Host() private parent: stripsBarComponent) {}
}
```

Then I created some tabs component which gets used this way:
```html
<tabs>
  <tab [selected]="false"></tab>
  <tab [selected]="true"></tab>
</tabs>
```

The tabs template uses the strip-bar and projects each tab within it:
```ts
@Component({
  template: `
    <div class="strips-bar__container">
      <strips-bar>
        <ng-content></ng-content>
      </strips-bar>
    </div>
  `
})
export Class TabsComponent {}
```

Each Tab uses the StripBarOptionDirective to register themselves with their parent StripBarComponent (found in the Tabs template):
```ts
@Component({
  template: `
    <div stripBarOption>
      <ng-content></ng-content>
    </div>
  `
})
export Class TabComponent {}
```

But because each tab is projected, the stripBarOptionDirective cannot get their parent StripBarComponent instance.

**TLDR;**

the developer types the following html:
```html
<tabs>
  <tab [selected]="false"></tab>
  <tab [selected]="true"></tab>
</tabs>
```

Which, once rendered looks like this:
```html
<tabs>
    <div class="nwa-strips-bar__container">
      <strips-bar>
        <div stripBarOption>
          <tab [selected]="false"></tab>
        </div>
        
        <div stripBarOption>
          <tab [selected]="true"></tab>
        </div>
     </strip-bar>
  </div>
</tabs>
```
But the stripBarOptionDirective cannot get the stripBarComponent event though they are wrapped in it.

**Comment by daicoden** [comment]
Ran into the same problem when using `<mat-form-field>`. Was trying to create a component which provides default behavior through template variables for a form field... same problem.

You can't easily build.

```
<app-my-form-field hint="A hint" hintToolTip = "more detailed rollover" label = "a label">
  <input matInput myDirective>
</app-my-form-field>

...
in my-form-field.component.html

<mat-form-field>
  <mat-label>{{label}}</mat-label>
  <app-ui-form-field-delegate> <!-- provided down below, but acts as the MatFormFieldControl for the MatFormfield -->
     <ng-content></ng-content>
  <app-ui-form-field-delegate>
  <mat-hint align="begin" *ngIf="hint">{{hint}}</mat-hint>
  <mat-hint align="end" *ngIf="hintToolTip">
   <mat-icon class="medium" [matTooltip]="hintToolTip" matTooltipClass="medium">help</mat-icon>
  </mat-hint>
  <mat-error>{{getErrors()}}</mat-error>
</mat-form-field>
```

In the example myDirective can't take in a MatFormField.

Would it be possible to specify an injector to ng-content, so you could at least manipulate what the content has access to?

----
```
delegate code

@Component({
  selector: 'app-ui-form-field-delegate',
  providers: [{provide: MatFormFieldControl, useExisting: UiFormFieldDelegateComponent}],
  template: `
    <ng-content></ng-content>`,
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class UiFormFieldDelegateComponent<T> implements MatFormFieldControl<T> {
  private _delegateFormField: MatFormFieldControl<T>;

  @Input()
  get delegateFormfield(): MatFormFieldControl<T> {
    return this._delegateFormField;
  }

  set delegateFormfield(val: MatFormFieldControl<T>) {
    this._delegateFormField = val;
    // workaround for https://github.com/angular/angular/issues/14935
    if (this.matFormField && this._delegateFormField instanceof AutocompleteComponent) {
      this._delegateFormField.matFormField = this.matFormField;
    }
  }

  constructor(@Optional() private matFormField: MatFormField) {
  }

  /**
   * MatFormFieldControl
   */
  get autofilled(): boolean {
    return this.delegateFormfield.autofilled;
  }

  /**
   * MatFormFieldControl
   */
  get controlType(): string {
    return this.delegateFormfield.controlType;
  }

  /**
   * MatFormFieldControl
   */
  get disabled(): boolean {
    return this.delegateFormfield.disabled;
  }

  /**
   * MatFormFieldControl
   */
  get empty(): boolean {
    return this.delegateFormfield.empty;
  }

  /**
   * MatFormFieldControl
   */
  get errorState(): boolean {
    return this.delegateFormfield.errorState;
  }

  /**
   * MatFormFieldControl
   */
  get focused(): boolean {
    return this.delegateFormfield.focused;
  }

  /**
   * MatFormFieldControl
   */
  get id(): string {
    return this.delegateFormfield.id;
  }

  /**
   * MatFormFieldControl
   */
  get ngControl(): NgControl | null {
    return this.delegateFormfield.ngControl;
  }

  /**
   * MatFormFieldControl
   */
  get placeholder(): string {
    return this.delegateFormfield.placeholder;
  }

  /**
   * MatFormFieldControl
   */
  get required(): boolean {
    return this.delegateFormfield.required;
  }

  /**
   * MatFormFieldControl
   */
  get shouldLabelFloat(): boolean {
    return this.delegateFormfield.shouldLabelFloat;
  }

  /**
   * MatFormFieldControl
   */
  get stateChanges(): Observable<void> {
    return this.delegateFormfield.stateChanges;
  }

  /**
   * MatFormFieldControl
   */
  get value(): T | null {
    return this.delegateFormfield.value;
  }

  /**
   * MatFormFieldControl
   */
  set value(v: T | null) {
    this.delegateFormfield.value = v;
  }

  /**
   * MatFormFieldControl
   */
  setDescribedByIds(ids: string[]): void {
    // Change after detection workaround
    setTimeout(() => this.delegateFormfield.setDescribedByIds(ids), 0);
  }

  /**
   * MatFormFieldControl
   */
  onContainerClick(event: MouseEvent): void {
    this.delegateFormfield.onContainerClick(event);
  }
}
```

**Comment by LastDragon-ru** [comment]
@daicoden, do you need something like? https://stackblitz.com/edit/components-issue-qygsj2

**Comment by andreElrico** [comment]
I would also like an option for the projected Content to be tied to the CD-Cycle of the projection host.
Almost 3 years ... is there any updates?

**Comment by waterplea** [comment]
It's unfortunate that we can utilize power of DI in dynamic content only if we use components :(

**Comment by crisbeto** [comment]
FYI we're reverting the feature for now in order to better figure out how the injector should behave. We'll target version 14 to re-release it.
